crates/ethereum/src/lib.rs

```diff
@@ -1,111 +1,295 @@
+        let res = self.http.post(self.url.clone()).json(&value).send().await?;
+
+        let status = res.status();
+        let (code, message) = (status.as_u16(), status.as_str());
+        if code != HTTP_OK {
+            tracing::error!(code, message, "Ethereum call failed");
+            anyhow::bail!(code);
+        }
+
+        let response: serde_json::Value = res.json().await?;
+        Ok(response["result"].clone())
+    }
+}
 
-    fn try_from(log: &ethers::types::Log) -> Result<Self, Self::Error> {
-        let block = BlockOrigin::try_from(log)?;
-        let transaction = TransactionOrigin::try_from(log)?;
-        let log_index = log.log_index.context("missing log index")?.as_u64();
+#[async_trait::async_trait]
+impl EthereumApi for EthereumClient {
+    async fn get_starknet_state(&self, address: &H160) -> anyhow::Result<EthereumStateUpdate> {
+        let hash = self.get_latest_block_hash().await?;
+        let hash = format!("0x{}", hex::encode(hash.as_bytes()));
+        let addr = format!("0x{}", hex::encode(address.as_bytes()));
+        Ok(EthereumStateUpdate {
+            global_root: self
+                .call_starknet_contract(&hash, &addr, "stateRoot()")
+                .await
+                .and_then(|value| get_h256(&value))
+                .and_then(get_felt)
+                .map(StateCommitment)?,
+            block_hash: self
+                .call_starknet_contract(&hash, &addr, "stateBlockHash()")
+                .await
+                .and_then(|value| get_h256(&value))
+                .and_then(get_felt)
+                .map(BlockHash)?,
+            block_number: self
+                .call_starknet_contract(&hash, &addr, "stateBlockNumber()")
+                .await
+                .and_then(|value| get_u256(&value))
+                .and_then(get_number)?,
+        })
+    }
 
-        Ok(EthOrigin {
-            block,
-            transaction,
-            log_index: EthereumLogIndex(log_index),
+    async fn get_chain(&self) -> anyhow::Result<EthereumChain> {
+        let id = self
+            .call_ethereum(serde_json::json!({
+                "jsonrpc": "2.0",
+                "method": "eth_chainId",
+                "params": [],
+                "id": 0
+            }))
+            .await
+            .and_then(|value| get_u256(&value))?;
+        Ok(match id {
+            x if x == U256::from(1u32) => EthereumChain::Mainnet,
+            x if x == U256::from(5u32) => EthereumChain::Goerli,
+            x => EthereumChain::Other(x),
         })
     }
 }
 
+fn encode_ethereum_call_data(signature: &[u8]) -> String {
+    let mut output: [u8; 32] = Default::default();
+    keccak_hash::keccak_256(signature, &mut output[..]);
+    format!("0x{}", hex::encode(&output[0..4]))
+}
+
+fn get_h256(value: &serde_json::Value) -> anyhow::Result<H256> {
+    use std::str::FromStr;
+    value
+        .as_str()
+        .and_then(|txt| H256::from_str(txt).ok())
```

Comment:
is non-fixed hex str length an issue here? ie. can the abi return a non 32 hex digit string and then does H256 support it?

Reply:
Why would it be?

https://docs.rs/primitive-types/latest/primitive_types/struct.H256.html#impl-FromStr-for-H256

https://docs.rs/fixed-hash/latest/src/fixed_hash/hash.rs.html#576

Reply:
Nice catch, thanks.

Done: https://github.com/eqlabs/pathfinder/pull/1110/commits/f6032a68e82d0ca1c1a8d35fbe71c360bfecf429

Reply:
Non-spec compliant eth servers can do whatever they wish. In which case this will error I think? Which is fine -- don't think we need to cater for short (or long) strings.